server/server.go

```diff
@@ -175,31 +188,46 @@ func (server *Server) startHTTPServers() {
 	server.serverEntryPoints = server.buildEntryPoints(server.globalConfiguration)
 
 	for newServerEntryPointName, newServerEntryPoint := range server.serverEntryPoints {
-		serverMiddlewares := []negroni.Handler{server.accessLoggerMiddleware, server.loggerMiddleware, metrics}
-		if server.globalConfiguration.Web != nil && server.globalConfiguration.Web.Metrics != nil {
-			if server.globalConfiguration.Web.Metrics.Prometheus != nil {
-				metricsMiddleware := middlewares.NewMetricsWrapper(middlewares.NewPrometheus(newServerEntryPointName, server.globalConfiguration.Web.Metrics.Prometheus))
-				serverMiddlewares = append(serverMiddlewares, metricsMiddleware)
-			}
-		}
-		if server.globalConfiguration.Web != nil && server.globalConfiguration.Web.Statistics != nil {
-			statsRecorder = middlewares.NewStatsRecorder(server.globalConfiguration.Web.Statistics.RecentErrors)
-			serverMiddlewares = append(serverMiddlewares, statsRecorder)
+		// build a filter chain of: log, metrics (wrapping prometheus), stats, authenticator, compress, router
+		// (working right to left)
+
+		var chain http.Handler = newServerEntryPoint.httpRouter
+
+		if server.globalConfiguration.EntryPoints[newServerEntryPointName].Compress {
+			chain = middlewares.NewCompress(chain)
 		}
+
 		if server.globalConfiguration.EntryPoints[newServerEntryPointName].Auth != nil {
-			authMiddleware, err := middlewares.NewAuthenticator(server.globalConfiguration.EntryPoints[newServerEntryPointName].Auth)
+			var err error
+			authConfig := server.globalConfiguration.EntryPoints[newServerEntryPointName].Auth
+			chain, err = middlewares.NewAuthenticator(authConfig, chain)
 			if err != nil {
 				log.Fatal("Error starting server: ", err)
 			}
-			serverMiddlewares = append(serverMiddlewares, authMiddleware)
 		}
-		if server.globalConfiguration.EntryPoints[newServerEntryPointName].Compress {
-			serverMiddlewares = append(serverMiddlewares, &middlewares.Compress{})
+
+		web := server.globalConfiguration.Web
+		if web != nil && web.Statistics != nil {
+			chain = middlewares.NewStatsRecorder(web.Statistics.RecentErrors, chain)
 		}
-		newsrv, err := server.prepareServer(newServerEntryPointName, newServerEntryPoint.httpRouter, server.globalConfiguration.EntryPoints[newServerEntryPointName], serverMiddlewares...)
+
+		if web != nil && web.Metrics != nil && web.Metrics.Prometheus != nil {
+			p := middlewares.NewPrometheus(newServerEntryPointName, web.Metrics.Prometheus)
+			chain = middlewares.NewMetricsWrapper(p, chain)
+		}
+
+		chain = common.NewAdapter(metrics, chain)
```

Comment:
AFAICS, this is the only occurrence where we need the negroni adapter. Do you think there's a way to get rid of this last bit (and that we should try to)?